community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/TemporalIndexCache.java

```diff
@@ -124,22 +142,22 @@ T select( ValueGroup valueGroup ) throws E
         switch ( valueGroup )
         {
         case DATE:
-            return date != null ? function.apply( date ) : orElse;
+            return parts[date.ordinal()] != null ? function.apply( parts[date.ordinal()] ) : orElse;
```

Comment:
sync in this class is wrong you have readers that will not gonna see parts that are created; Instead of doing this kind of error-prone synchronization would recommend available sync collections. For example: CHM

Reply:
What is CHM?

Reply:
CHM -> ConcurrentHashMap

Reply:
The reason we didn't go for that is the we wanted to ensure that `factory.newLocalDateTime()` (and similar) is only called once. And you should only pass side-effect-free things to `ConcurrentHashMap.computeIfAbsent` thus it can't guarantee that its only called once, if we don't add synchronization anyways. Is there an error in this logic?